### PR TITLE
feat(expr): add `jsonb_populate_record(set)` function

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds the `jsonb_populate_record` function and its set-returning form:

```
jsonb_populate_record ( base anyelement, from_json jsonb ) → anyelement
jsonb_populate_recordset ( base anyelement, from_json jsonb ) → setof anyelement
```

To support `jsonb_populate_recordset`, some improvements have been made to the `#[function]` macro, in order to support `Option` inputs and the `Context` argument for table functions.

### Remaining issue

Calling the scalar function `jsonb_populate_record` in the FROM clause triggers a panic in the planner.

```sql
dev=> select * from jsonb_populate_record(
    null::struct<a int, b int>,
    '{"a": 1, "b": 2}' 
);
ERROR:  Panicked when processing: insert at index 1 exceeds fixbitset size 1
```

Although calling it in the SELECT list is normal, it seems that the above form is more commonly used.

```sql
dev=> select jsonb_populate_record(
    null::struct<a int, b int>,
    '{"a": 1, "b": 2}'
);
 jsonb_populate_record 
-----------------------
 (1,2)
```

---

This PR also implements a similar function `jsonb_to_record(set)` on the backend. However, they are not enabled due to lack of frontend support.

These functions are designed to be called in the FROM clause, with type defined by an AS clause.

```sql
postgres=# select * from jsonb_to_record('{"a":1,"b":"foo","c":"bar"}')
    as x(a int, b text, d text);
 a |  b  | d 
---+-----+---
 1 | foo | 
```

It seems that we don't support such usage in both parser and frontend.

```sql
-- risingwave
dev=> select * from jsonb_to_record('{"a":1,"b":"foo","c":"bar"}')
    as x(a int, b text, d text);
ERROR:  QueryError

Caused by:
  sql parser error: Expected ), found: int at line:2, column:15
Near "
    as x(a"
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Add 2 functions:
```
jsonb_populate_record ( base anyelement, from_json jsonb ) → anyelement
jsonb_populate_recordset ( base anyelement, from_json jsonb ) → setof anyelement
```

See [postgres documents](https://www.postgresql.org/docs/current/functions-json.html) for detailed descriptions.

Note that we have different usages compared to pg.

```sql
-- risingwave
select (jsonb_populate_record(
    null::struct<a int, b text[], c struct<d int, e text>>,
    '{"a": 1, "b": ["2", "a b"], "c": {"d": 4, "e": "a b c"}, "x": "foo"}'
)).*;
----
1 {2,"a b"} (4,"a b c")

select * from jsonb_populate_recordset(
    null::struct<a int, b int>,
    '[{"a":1,"b":2}, {"a":3,"b":4}]'::jsonb
);
----
1 2
3 4
```

In Postgres, users need to define a composite type using `CREATE TYPE` before calling these functions.
In RisingWave, users should use inline `struct` type instead.

### CODE REVIEW

**Comment by fuyufjh** [comment]
Regarding of the remaining issue: I think it's okay to keep the status quo i.e. do not support this usage now. I can't see any real benefit or use cases. What do you think? cc. @xiangjinwu  

------

Did some quick experiments and here is the result:

```
test=# select md5('hello');
               md5
----------------------------------
 5d41402abc4b2a76b9719d911017c592
(1 row)

test=# select * from md5('hello');
               md5
----------------------------------
 5d41402abc4b2a76b9719d911017c592
(1 row) 
```

While operators can't

```
test=# select * from 1+1;
ERROR:  syntax error at or near "1"
```

**Comment by wangrunji0408** [comment]
`select * from a_scalar_function_that_returns_record()` seems to be useful.
But we support an equivalent way: `select (a_scalar_function_that_returns_record()).*;`
So I also agree to leave this issue to the future.

**Comment by xiangjinwu** [comment]
> Regarding of the remaining issue: I think it's okay to keep the status quo i.e. do not support this usage now. I can't see any real benefit or use cases. What do you think? cc. @xiangjinwu

Its major usage is to convert from a loosely typed `jsonb` to strongly typed `struct`, as we have discussed on the `map` https://github.com/risingwavelabs/risingwave/issues/13387#issuecomment-1809644378 . According to [PostgreSQL](https://www.postgresql.org/docs/16/functions-json.html):
> typical use would be to reference a `json` or `jsonb` column laterally from another table in the query's `FROM` clause. Writing `json_populate_record` in the `FROM` clause is good practice, since all of the extracted columns are available for use without duplicate function calls.

I will give an example on how it could help if we decided to use `jsonb` for proto `<string, Location>`:
```
select (lookup(col1, 'SFO')).latitude from t; -- if we had a `lookup` function on native `map` or `struct[]`
select col1 -> 'SFO' ->> 'latitude' from t; -- if we used `jsonb`

select
  latitude
from
  t,
  json_populate_record(null::struct<latitude float8, longitude float8>, col1 -> 'SFO');
-- if we used `jsonb`, and `json_populate_record` helps use convert inner `jsonb` back to strongly typed struct `Location`.
```

It is true that we can do `(...).*` in select-list as well. But the **lateral-from** syntax is more convenient if we want to process further immediately, for example `select tand(latitude) * sind(longitude) from ...`

That being said, it is always possible to implement a functionality that can only be used in limited cases, and then improved in future iterations.

**Review by TennyZhuang — APPROVED** [review]
LGTM

File: src/common/src/types/jsonb.rs [hunk 1]
```diff
@@ -380,6 +383,96 @@ impl<'a> JsonbRef<'a> {
...
+        let object = self.0.as_object().ok_or_else(|| {
+            format!(
+                "cannot call populate_composite on a jsonb {}",
+                self.type_name()
+            )
+        })?;
+        let mut fields = Vec::with_capacity(ty.len());
+        for ((name, ty), base_field) in ty.iter().zip_eq_debug(base.iter_fields_ref()) {
+            let datum = match object.get(name) {
+                Some(v) => Self(v).to_datum(ty)?,
```
Comment (xiangjinwu) [inline src/common/src/types/jsonb.rs:468]: Is the fallback-to-base behavior supposed to happen recursively? I just tried the following in PostgreSQL:
```
test=# create table pt (x int, y int);
CREATE TABLE
test=# create table sg (a pt, b pt);
CREATE TABLE
test=# select jsonb_populate_record(row(row(1, 2)::pt, row(3, 4)::pt)::sg, '{"a": {"x":9},"b":{"y":17}}');
 jsonb_populate_record 
-----------------------
 ("(9,2)","(3,17)")
```

But in this PR:
```
dev=> select jsonb_populate_record(row(row(1, 2), row(3, 4))::struct<a struct<x int, y int>, b struct<x int, y int>>, '{"a": {"x":9},"b":{"y":17}}');
 jsonb_populate_record 
-----------------------
 ("(9,)","(,17)")
(1 row)
```

**Review by xiangjinwu — APPROVED** [review]
Just wondering the following:
* How is frontend agg name `jsonb_populate_recordset` mapped to the proto enum, by the macro?
* How does frontend infer the return type is same as first argument for `jsonb_populate_record`?

**Comment by wangrunji0408** [comment]
> How is frontend agg name `jsonb_populate_recordset` mapped to the proto enum, by the macro?

by the `PbType::from_str_name` function generated by prost:
https://github.com/risingwavelabs/risingwave/blob/86ac793d95023614c9b5e441b66a4753f924082c/src/prost/src/lib.rs#L171-L177

> How does frontend infer the return type is same as first argument for `jsonb_populate_record`?

The `#[function]` macro generates a type inference function for this function.

https://github.com/risingwavelabs/risingwave/blob/86ac793d95023614c9b5e441b66a4753f924082c/src/expr/macro/src/gen.rs#L113-L117

and frontend will call this function at the end of `infer_type`.

https://github.com/risingwavelabs/risingwave/blob/86ac793d95023614c9b5e441b66a4753f924082c/src/frontend/src/expr/type_inference/func.rs#L72

File: src/common/src/types/jsonb.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/common/src/types/jsonb.rs:468]: It seems so. I should fix this.

**Pull request merged**